### ISSUE TITLE
Leaky execution related to a race condition when a listener is added to a MemoryStream

### ISSUE THREAD

**Issue by whitecolor** [staltz/xstream#53]
http://www.webpackbin.com/Ny3m5bc4-

See console output.

There is three ending listeners attached, one of the them (in the middle) takes 0 (or 1 item), and this causes kind of a leak after all all listeners detached
.

**Comment by whitecolor** [comment]
Not pointing on urgency but commenting the `weight (could)` - this issue needs to be checked and fixed because it may mean that there is something wrong with whole approach to inner cycles detection.

**Comment by Hypnosphi** [comment]
In fact, this bug has nothing to do with `imitate` and can be simplified to this: http://www.webpackbin.com/Vywzoz0E-

the problem here is the combination of MemoryStream with synchronous starting value and `take(1)`.

Let's take a look at [`MemoryStream#_add` method](https://github.com/staltz/xstream/blob/master/src/core.ts#L1777)
```js
_add(il: InternalListener<T>): void {
    if (this._has) { il._n(this._v); }
    super._add(il);
}
```

Here, if a value is already present, it is emitted before the listener is really added. In case of `take(1)` it (https://github.com/staltz/xstream/blob/master/src/core.ts#L892)[leads to instant completion] of outer stream. TakeOperator therefore stops and tries to remove itself from the `source$` MemoryStream internal listeners list. But it isn't present at that list yet, so it's a no-op. then, in the second line of `MemoryStream#_add`, it finally adds itself and stays there forever.

A simple solution would be if MemoryStream just added listeners before emitting the current value, but I'm not sure it wouldn't lead to other bugs, so maybe a better option would be if operators removed itself asynchronously.

**Comment by staltz** [comment]
Fixed in v7.0
